doc/cabal-package-description-file.rst

```diff
@@ -318,24 +318,39 @@ describe the package as a whole:
     tools require the package-name specified for this field to match
     the package description's file-name :file:`{package-name}.cabal`.
 
-    Package names are case-sensitive and must match the regular expression
-    (i.e. alphanumeric "words" separated by dashes; each alphanumeric
-    word must contain at least one letter):
-    ``[[:digit:]]*[[:alpha:]][[:alnum:]]*(-[[:digit:]]*[[:alpha:]][[:alnum:]]*)*``.
+    A valid package name comprises an alphanumeric 'word'; or two or more
+    such words separated by a hyphen character (``-``). A word cannot be
+    comprised only of the digits ``0`` to ``9``.
 
-    Or, expressed in ABNF_:
+    An alphanumeric character belongs to one of the Unicode Letter categories
+    (Lu (uppercase), Ll (lowercase), Lt (titlecase), Lm (modifier), or
+    Lo (other)) or Number categories (Nd (decimal), Nl (letter), or No (other)).
+
+    Package names are case-sensitive.
+
+    Expressed as a regular expression:
+
+    ``[0-9]*[\p{L}\p{N}-[0-9]][\p{L}\p{N}]*(-[0-9]*[\p{L}\p{N}-[0-9]][\p{L}\p{N}]*)*``
+
+    Expressed in ABNF_:
 
     .. code-block:: abnf
 
         package-name      = package-name-part *("-" package-name-part)
-        package-name-part = *DIGIT UALPHA *UALNUM
+        package-name-part = *DIGIT UALPHANUM-NOT-DIGIT *UALNUM
 
-        UALNUM = UALPHA / DIGIT
-        UALPHA = ... ; set of alphabetic Unicode code-points
+        DIGIT = %x30-39 ; 0-9
+
+        UALNUM = UALPHANUM-NOT-DIGIT / DIGIT
+        UALPHANUM-NOT-DIGIT = ... ; set of Unicode code-points in Letter or
+                                  ; Number categories, other than the DIGIT
+                                  ; code-points
 
     .. note::
 
-        Hackage restricts package names to the ASCII subset.
+        Hackage will not accept package names that use alphanumeric characters
+        other than ``A`` to ``Z``, ``a`` to ``z``, and ``0`` to ``9``
```

Comment:
It does accept dashes though. Not sure about underscores.

Reply:
Dash (U+002D; hyphen-minus) is not an alphanumeric character (it is in Unicode category Pd), so I think the statement is accurate. The note is contrasting with what Cabal will accept.

Underscore (U+005F; low line) is in category Pc. It is not accepted by Cabal. I understand from a discussion on Matrix that Hackage uses the Cabal parser and then overlays an ASCII-only test, so it should not be accepted by Hackage as well.

Reply:
BTW, arguably out of scope of this PR, but while we are talking about Hackage, perhaps it's worth pointing out that `cabal check` would warn (or error) at such names? If that's true ATM. @ffaf1?

This would tie the remark back to cabal, of which this is the documentation. Otherwise, why not mention what Stackage accepts?

Reply:
`check` does check for non-ASCII in custom fields and flags but not on package name.

Since it is a check that can be performed locally, it should be added to `cabal check`, thanks for bringing this to my attention.

Reply:
@Mikolaj, on what Stackage accepts, Stackage describes itself as a subset of packages from Hackage, so I don't think it needs a separate mention.